demux/demux_playlist.c

```diff
@@ -407,8 +410,13 @@ static bool scan_dir(struct pl_parser *p, char *path,
     if (dir_entries)
         qsort(dir_entries, num_dir_entries, sizeof(dir_entries[0]), cmp_dir_entry);
 
+    struct MPOpts *opts = mp_get_config_group(NULL, p->global, &mp_opt_root);
+    bool recursive = dir_mode == DIR_RECURSIVE ||
+                     (dir_mode == DIR_AUTO && opts->shuffle);
+    talloc_free(opts);
```

Comment:
That happens for each directory that gets loaded, which could be a ton of times for recursive loading. Can we pull that out so that it only happens once?

`parse_dir()` seems like a good place, and then pass it in as a parameter.
Or, and I'm not 100% sure that's fine to do but I think it is, we could set `p.opts.dir_mode` to recursive in `parse_dir()` if it's set to auto.

Reply:
Moved to `parse_dir()`. Overwriting `p.opts.dir_mode` would change the option value without user interaction which wm4 didn't want (aa5f234b439 d07b7f068db). That code blocks runs in 50 nanoseconds anyway.

Reply:
Would it? I think the options we get from `mp_get_config_group()` are a copy of the "real" options.

Reply:
Alternative idea if you really want to avoid modifying the `p->opts->dir_mode`, since we already pass in a new parameter to `scan_dir()` we could make it a `dir_mode` that lives on the stack of `parse_dir()` and reflects the value of the option, unless that option is auto, in which case it gets set to recursive or lazy.

Reply:
Good point, I updated it that way.